java/src/shared/model/game/Game.java

```diff
@@ -704,8 +706,19 @@ public void buildCity(int playerIndex, VertexLocation vertex) throws InvalidPlay
      */
     @Override
     public void discardCards(int playerIndex, List<ResourceType> cards) throws PlayerExistsException, InsufficientResourcesException, InvalidTypeException {
-        if(canDiscardCards(playerIndex) && this.turnTracker.canDiscard()) {
+        if(canDiscardCards(playerIndex) && this.turnTracker.canDiscard() && needToDiscard.containsKey(playerIndex)) {
             playerManager.discardResourceType(playerIndex, cards);
+            needToDiscard.put(playerIndex,false);
+        }
+        boolean change = true;
+        for(HashMap.Entry<Integer,Boolean> entry : needToDiscard.entrySet()){
```

Comment:
Is the map needToDiscard used to make sure a player only discards once? The players have a data member that keeps track of whether or not they have discarded this turn. I think we should leave the rollNumber method the way it is and this method should just ask every player if they can discard and whether or not they have discarded yet. Once everyone has discarded move the phase to robbing.